src/main/java/com/powsybl/balances_adjustment/util/ControlArea.java

```diff
@@ -0,0 +1,109 @@
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import com.powsybl.cgmes.extensions.CgmesControlAreas;
+import com.powsybl.commons.PowsyblException;
+import com.powsybl.iidm.network.*;
+import org.jgrapht.Graph;
+import org.jgrapht.alg.connectivity.ConnectivityInspector;
+import org.jgrapht.graph.Pseudograph;
+
+/**
+ * @author Marcos de Miguel <demiguelm at aia.es>
+ */
+public class ControlArea implements NetworkArea {
+
+    private final Set<Object> terminalsAndBoundaries = new HashSet<>();
+    private final Set<String> voltageLevelIdsCache;
+
+    public ControlArea(Network network, String controlAreaId) {
+        CgmesControlAreas cgmesControlAreaMapping = network.getExtension(CgmesControlAreas.class);
+        terminalsAndBoundaries.addAll(cgmesControlAreaMapping.getCgmesControlArea(controlAreaId).getTerminals());
+        terminalsAndBoundaries.addAll(cgmesControlAreaMapping.getCgmesControlArea(controlAreaId).getBoundaries());
+
+        voltageLevelIdsCache = getVoltageLevelIds(controlAreaId, network, terminalsAndBoundaries);
+    }
+
+    @Override
+    public double getNetPosition() {
+        return terminalsAndBoundaries.parallelStream().mapToDouble(o -> {
+            if (o instanceof Terminal) {
+                return getLeavingFlow((Terminal) o);
+            } else if (o instanceof Boundary) {
+                return -((Boundary) o).getP();
+            }
+            throw new AssertionError();
+        }).sum();
+    }
+
+    @Override
+    public Collection<String> getVoltageLevelIds() {
+        return Collections.unmodifiableSet(voltageLevelIdsCache);
+    }
+
+    private static Set<String> getVoltageLevelIds(String controlAreaId, Network network, Set<Object> terminalsAndBoundaries) {
+        Graph<Identifiable, Object> voltageLevelGraph = createVoltageLevelGraph(network, terminalsAndBoundaries);
+        List<Set<Identifiable>> connectedSets = new ConnectivityInspector<>(voltageLevelGraph)
+                .connectedSets()
+                .stream()
+                .filter(set -> terminalsAndBoundaries.stream().anyMatch(o -> set.contains(getVoltageLevel(o))))
+                .collect(Collectors.toList());
+        if (connectedSets.isEmpty()) {
+            throw new PowsyblException("Control area " + controlAreaId + " does not contain any voltage level");
+        }
+        if (connectedSets.size() > 1) {
+            throw new PowsyblException("Control area " + controlAreaId + " contains several connected sets of voltage levels");
+        }
+        return connectedSets.iterator().next()
+                .stream()
+                .filter(o -> o instanceof VoltageLevel)
+                .map(Identifiable::getId)
+                .collect(Collectors.toSet());
+    }
+
+    private static Graph<Identifiable, Object> createVoltageLevelGraph(Network network, Set<Object> terminalsAndBoundaries) {
+        Graph<Identifiable, Object> voltageLevelGraph = new Pseudograph<>(Identifiable.class);
```

Comment:
It's the class for edges which is expected as argument: `new Pseudograph<>(Object.class)`

Reply:
Oops, yes, I think I did that by mistake!